modules/optional-enabling-read-only-mode-backup-restore.adoc

```diff
@@ -0,0 +1,357 @@
+
+* You have cloned the `https://github.com/quay/quay.git` repository.
+
+[id="creating-service-keys"]
+== Creating service keys 
+
+{productname} uses service keys to communicate with various components. These keys are used to sign completed requests, such as requesting to scan images, login, storage access, and so on.
+
+.Procedure
+
+. Change into the directory of your {productname} deployment and create a virtual environment inside of that directory:
++
+[source,terminal]
+----
+$ cd <$QUAY>/quay && virtualenv -v venv
+----
+
+. Activate the virtual environment by entering the following command:
++
+[source,terminal]
+----
+$ source venv/bin/activate
+----
+
+. Optional. Install the `pip` CLI tool if you do not have it installed:
++
+[source,terminal]
+----
+$ venv/bin/pip install --upgrade pip
+----
+
+. In your {productname} directory, create a `requirements-generatekeys.txt` file with the following content:
++
+[source,terminal]
+----
+$ cat << EOF > requirements-generatekeys.txt
+cryptography==3.4.7
+pycparser==2.19
+pycryptodome==3.9.4
+pycryptodomex==3.9.4
+pyjwkest==1.4.2
+PyJWT==1.7.1
+Authlib==1.0.0a2
+EOF
+----
+
+. Enter the following command to install the Python dependencies defined in the `requirements-generatekeys.txt` file:
++
+[source,terminal]
+----
+$ venv/bin/pip install -r requirements-generatekeys.txt
+----
+
+. Enter the following command to create the necessary service keys:
++
+[source,terminal]
+----
+$ PYTHONPATH=. venv/bin/python /<path_to_cloned_repo>/tools/generatekeypair.py quay-readonly
+----
++
+Example output
++
+[source,terminal]
+----
+Writing public key to quay-readonly.jwk
+Writing key ID to quay-readonly.kid
+Writing private key to quay-readonly.pem
+----
+
+. Enter the following command to deactivate the virtual environment:
++
+[source,terminal]
+----
+$ deactivate
+----
+
+[id="adding-keys-postgresql-database"]
+== Adding keys to the PostgreSQL database
+
+Use the following procedure to add your service keys to the PostgreSQL database.
+
+.Prerequistes
+
+* You have created the service keys.
+
+.Procedure
+
+. Enter the following command to enter your {productname} database environment:
++
+[source,terminal]
+----
+$ podman exec -it postgresql-quay psql -U postgres -d quay
+----
+
+. Add the service key to your {productname} database by entering the following query:
++
+[source,terminal]
+----
+quay=# INSERT INTO servicekey 
+  ('name', 'service', 'metadata', 'kid', 'jwk', 'expiration_date')
+  VALUES ("quay-readonly",
+           "quay",
+           "{}",
+           "{<contents_of_.kid_file>}",
+           "{<contents_of_.jwk_file>}", 
+           "{<expiration_date_of_read-only>}");
+----
++
+Example output
++
+[source,terminal]
+----
+INSERT 0 1
+----
+
+. Next, add the key approval with the following query:
++
+[source,terminal]
+----
+quay=# INSERT INTO servicekeyapproval ('approval_type', 'notes')
+  VALUES ("Super User API",
+           {include_notes_here_on_why_this_is_being_added});
+----
```

Comment:
In Quay 3.11 the SQL Command is not correct, the name of approval_type should be "ServiceKeyApprovalType.SUPERUSER", here also give example of the command.

```
quayregistry-quay-database=# select * from servicekeyapproval;
 id | approver_id |          approval_type           |       approved_date        | notes 
----+-------------+----------------------------------+----------------------------+-------
  1 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 03:47:48.181347 | 
  2 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 03:47:55.808087 | 
  3 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 03:49:04.27095  | 
  4 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 03:49:05.46235  | 
  5 |           1 | ServiceKeyApprovalType.SUPERUSER | 2024-05-07 04:05:10.296796 | 
  6 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 04:43:57.099776 | 
  7 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 04:43:58.199101 | 
  8 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 04:43:58.740628 | 
  9 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 04:44:57.782562 | 
 10 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:04:46.102356 | 
 11 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:04:46.907342 | 
 12 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:15:00.004144 | 
 13 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:15:00.288843 | 
 14 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:26:13.871902 | 
 15 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:26:44.553869 | 
 16 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:26:44.980913 | 
 17 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:27:15.044647 | 
 18 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:46:41.631166 | 
 19 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:47:12.69024  | 
 20 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:47:12.72553  | 
 21 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 06:47:53.351615 | 
 22 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 07:04:05.20755  | 
 23 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 07:04:35.534945 | 
 24 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 07:04:36.499352 | 
 25 |             | ServiceKeyApprovalType.AUTOMATIC | 2024-05-07 07:05:05.599616 | 
(25 rows)
```

```
INSERT INTO servicekey (name, service, metadata, kid, jwk, created_date, expiration_date) 
  VALUES ('quay-readonly', 
                     'quay', 
                     '{}', 
                     'KVaF5mYCwo0iY1UTkq7Pl91oLE_yT9Upyo5Vl1Pb1K0', 
                     '{"n":"l5hXx6xvOL44uBI9uFtl0sOQclcyj7pLDXFdPN7kKUuuXa4Dw8L7ah0fPynNFN_0_KaE6xQ50nOpEcQSIspj0kOl2Nzpis2nWUr0-aNeWB87VlMsbNf5ZvrE36AGNdWVAi0pZucQhCjs5x0vQx2eG9o7X66frN8LnBjN9e71oH6Fo_oUIIwe4rT8ysBhOfri2qKs3AMIdDeB9uJ4E1-1CDax_bmzW7ERD1t8Pb2uC-FwsHvUHBI0myZCLyqu1o0kk6sbFJMMfufs1oSe4xrGjFmV77CMJMGrxPAAdNshd4Fy5-8xqr04LqOMvP7rGiIMU3swhgeexz911ybAQnbLfQ","e":"AQAB","kty":"RSA","kid":"KVaF5mYCwo0iY1UTkq7Pl91oLE_yT9Upyo5Vl1Pb1K0"}', 
                     '2023-06-10 12:10:10', 
                     '2024-06-10 12:10:10');
```